### Logistic regression ###

First, we replace all values in the raw data with their corresponding WoE values. We then perform logistic regression on each variable to obtain the coefficients needed for a credit score card.

In [1]:
import pandas as pd
import numpy as np
from sklearn.linear_model import LogisticRegression

In [2]:
predictions_df = pd.read_csv('../../data/cs-training.csv')

In [3]:
# The original data.
predictions_df.head()

Unnamed: 0  SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0           1                 1                              0.766127   45   
1           2                 0                              0.957151   40   
2           3                 0                              0.658180   38   
3           4                 0                              0.233810   30   
4           5                 0                              0.907239   49   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                     2   0.802982         9120.0   
1                                     0   0.121876         2600.0   
2                                     1   0.085113         3042.0   
3                                     0   0.036050         3300.0   
4                                     1   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                               13                        0   
1                                4                        0   
2                                2                        1   
3                                5                        0   
4                                7                        0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                             6                                     0   
1                             0                                     0   
2                             0                                     0   
3                             0                                     0   
4                             1                                     0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [4]:
predictions_df['SeriousDlqin2yrs'].value_counts()

0    139974
1     10026
Name: SeriousDlqin2yrs, dtype: int64

In [5]:
np.log(139974/10026)

2.636274972235245

When we do logistic regression, we will expect the coefficient to be about 1 and the intercept to be about 2.64.

In [6]:
woe_df = pd.read_csv('../../data/report_woe.csv')

In [7]:
woe_df_nan = woe_df[woe_df['QuantileRange'].isna()]

In [8]:
# Table of correspodning WoE for NaN values.
woe_df_nan

Unnamed: 0                              Variable QuantileRange  \
4            4  RevolvingUtilizationOfUnsecuredLines           NaN   
10           5                                   age           NaN   
13           2  NumberOfTime30-59DaysPastDueNotWorse           NaN   
18           4                             DebtRatio           NaN   
22           3                         MonthlyIncome           NaN   
30           4               NumberOfTimes90DaysLate           NaN   
33           2          NumberRealEstateLoansOrLines           NaN   
37           3  NumberOfTime60-89DaysPastDueNotWorse           NaN   
40           2                    NumberOfDependents           NaN   

    EventsCount  EventsSum  % of Qualified  DelinquentCount  % of Delinquent  \
4         22602       2481        0.247457            20121         0.143748   
10         9536        764        0.076202             8772         0.062669   
13        20631       3628        0.361859            17003         0.121473   
18         4134        340        0.033912             3794         0.027105   
22        35592       1962        0.195691            33630         0.240259   
30          139         88        0.008777               51         0.000364   
33       114826       7842        0.782166           106984         0.764313   
37         1118        561        0.055955              557         0.003979   
40        23446       1763        0.175843            21683         0.154907   

         WoE        IV  
4  -0.543173  0.056332  
10 -0.195523  0.002646  
13 -1.091567  0.262398  
18 -0.224044  0.001525  
22  0.205179  0.009144  
30 -3.181786  0.026768  
33 -0.023090  0.000412  
37 -2.643431  0.137393  
40 -0.126763  0.002654

In [9]:
woe_df_not_nan = woe_df[woe_df['QuantileRange'].notna()]

In [10]:
# Table of correspodning WoE for non-NaN values.
woe_df_not_nan

Unnamed: 0                              Variable     QuantileRange  \
0            0  RevolvingUtilizationOfUnsecuredLines       (0.0, 0.49]   
1            1  RevolvingUtilizationOfUnsecuredLines       (0.5, 0.99]   
2            2  RevolvingUtilizationOfUnsecuredLines       (1.0, 1.49]   
3            3  RevolvingUtilizationOfUnsecuredLines    (1.5, 60000.0]   
5            0                                   age          (18, 25]   
6            1                                   age          (26, 34]   
7            2                                   age          (35, 44]   
8            3                                   age          (45, 64]   
9            4                                   age         (65, 120]   
11           0  NumberOfTime30-59DaysPastDueNotWorse    (-0.098, 0.99]   
12           1  NumberOfTime30-59DaysPastDueNotWorse      (2.0, 100.0]   
14           0                             DebtRatio    (0.0, 0.31249]   
15           1                             DebtRatio  (0.3125, 0.6249]   
16           2                             DebtRatio     (0.625, 2.49]   
17           3                             DebtRatio   (2.5, 150000.0]   
19           0                         MonthlyIncome         (0, 4999]   
20           1                         MonthlyIncome  (10000, 1000000]   
21           2                         MonthlyIncome      (5000, 9999]   
23           0       NumberOfOpenCreditLinesAndLoans   (-0.058, 11.59]   
24           1       NumberOfOpenCreditLinesAndLoans     (11.6, 17.39]   
25           2       NumberOfOpenCreditLinesAndLoans      (17.4, 58.0]   
26           0               NumberOfTimes90DaysLate     (-0.098, 1.0]   
27           1               NumberOfTimes90DaysLate        (1.5, 4.9]   
28           2               NumberOfTimes90DaysLate      (10.0, 98.0]   
29           3               NumberOfTimes90DaysLate        (5.0, 9.9]   
31           0          NumberRealEstateLoansOrLines        (1.0, 2.9]   
32           1          NumberRealEstateLoansOrLines       (3.0, 54.0]   
34           0  NumberOfTime60-89DaysPastDueNotWorse      (-10.0, 0.0]   
35           1  NumberOfTime60-89DaysPastDueNotWorse       (0.01, 1.0]   
36           2  NumberOfTime60-89DaysPastDueNotWorse      (2.0, 100.0]   
38           0                    NumberOfDependents      (-0.02, 1.0]   
39           1                    NumberOfDependents       (2.0, 20.0]   

    EventsCount  EventsSum  % of Qualified  DelinquentCount  % of Delinquent  \
0         97178       2620        0.261321            94558         0.675540   
1         26905       3689        0.367943            23216         0.165859   
2          2715       1080        0.107720             1635         0.011681   
3           600        156        0.015560              444         0.003172   
5          3027        338        0.033712             2689         0.019211   
6         15019       1665        0.166068            13354         0.095403   
7         26317       2344        0.233792            23973         0.171268   
8         67502       4225        0.421404            63277         0.452063   
9         28599        690        0.068821            27909         0.199387   
11       126018       5041        0.502793           120977         0.864282   
12         3351       1357        0.135348             1994         0.014246   
14        60919       3455        0.344604            57464         0.410533   
15        38557       2745        0.273788            35812         0.255848   
16        15741       1778        0.177339            13963         0.099754   
17        30649       1708        0.170357            28941         0.206760   
19        51468       4584        0.457211            46884         0.334948   
20        18315        792        0.078995            17523         0.125188   
21        44625       2688        0.268103            41937         0.299606   
23       115311       7738        0.77179

In [11]:
predictions_df = predictions_df.drop(columns='Unnamed: 0')

In [12]:
predictions_df.head()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0                 1                              0.766127   45   
1                 0                              0.957151   40   
2                 0                              0.658180   38   
3                 0                              0.233810   30   
4                 0                              0.907239   49   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                     2   0.802982         9120.0   
1                                     0   0.121876         2600.0   
2                                     1   0.085113         3042.0   
3                                     0   0.036050         3300.0   
4                                     1   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                               13                        0   
1                                4                        0   
2                                2                        1   
3                                5                        0   
4                                7                        0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                             6                                     0   
1                             0                                     0   
2                             0                                     0   
3                             0                                     0   
4                             1                                     0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [13]:
# We will create a separate Dataframe for the WoE values.
predictions_df_woe = predictions_df.copy()

##### WoE for non-NaN values #####

In [14]:
# List of all variables in the data.
var_list = list(woe_df_not_nan['Variable'].unique())

In [15]:
# Function to extract each range of values for a given variable and the WoE for that range as a tuple.
def make_map(ref_df, variable):
    tuple_list = []
    for item in ref_df[(ref_df['Variable'] == variable) & ref_df['QuantileRange'].notna()].iterrows():
        range_bounds = item[1]['QuantileRange'][1:-1].split(', ')
        range_bounds = [float(num) for num in range_bounds]
        tuple_list.append((range_bounds, item[1]['WoE']))
    return tuple_list

In [16]:
# Use the make_map function to extract the range-to-WoE map for each variable and store all the maps in a dictionary.
maps = dict()
for var in var_list:
    maps[var] = (make_map(woe_df, var))

In [17]:
predictions_df_woe.head()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines  age  \
0                 1                              0.766127   45   
1                 0                              0.957151   40   
2                 0                              0.658180   38   
3                 0                              0.233810   30   
4                 0                              0.907239   49   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                                     2   0.802982         9120.0   
1                                     0   0.121876         2600.0   
2                                     1   0.085113         3042.0   
3                                     0   0.036050         3300.0   
4                                     1   0.024926        63588.0   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                               13                        0   
1                                4                        0   
2                                2                        1   
3                                5                        0   
4                                7                        0   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                             6                                     0   
1                             0                                     0   
2                             0                                     0   
3                             0                                     0   
4                             1                                     0   

   NumberOfDependents  
0                 2.0  
1                 1.0  
2                 0.0  
3                 0.0  
4                 0.0

In [18]:
# Function that will be used to replace each value in a column with its corresponding WoE.
def woe_replace(in_num, tuples):
    for tup in tuples:
        if tup[0][0] <= in_num <= tup[0][1]:
            return tup[1]

In [19]:
# Apply the woe_replace function to each column of the original data.
for var in var_list:
    predictions_df_woe[var] = predictions_df_woe[var].apply(woe_replace, args=(maps[var],))

In [20]:
predictions_df_woe.head()

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines       age  \
0                 1                             -0.796789  0.070228   
1                 0                             -0.796789 -0.311206   
2                 0                             -0.796789 -0.311206   
3                 0                              0.949764 -0.554284   
4                 0                             -0.796789  0.070228   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                             -2.251409  -0.575353       0.111096   
1                              0.541721   0.175061      -0.311170   
2                                   NaN   0.175061      -0.311170   
3                              0.541721   0.175061      -0.311170   
4                                   NaN   0.175061       0.460433   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                         0.035635                 0.176674   
1                        -0.004249                 0.176674   
2                        -0.004249                 0.176674   
3                        -0.004249                 0.176674   
4                        -0.004249                 0.176674   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                     -0.592980                              0.288208   
1                           NaN                              0.288208   
2                           NaN                              0.288208   
3                           NaN                              0.288208   
4                      0.188639                              0.288208   

   NumberOfDependents  
0           -0.352272  
1            0.078749  
2            0.078749  
3            0.078749  
4            0.078749

In [21]:
##### WoE for NaN values #####

In [22]:
# List of all variables that have NaNs
var_list_null = list(woe_df_nan['Variable'].unique())

In [23]:
# For each variable, replace all NaN values with the appropriate WoE
for var in var_list_null:
    predictions_df_woe[var].fillna(woe_df_nan[(woe_df_nan['Variable']==var)].reset_index().loc[0, 'WoE'], inplace=True)

In [24]:
predictions_df_woe.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 150000 entries, 0 to 149999
Data columns (total 11 columns):
 #   Column                                Non-Null Count   Dtype  
---  ------                                --------------   -----  
 0   SeriousDlqin2yrs                      150000 non-null  int64  
 1   RevolvingUtilizationOfUnsecuredLines  150000 non-null  float64
 2   age                                   150000 non-null  float64
 3   NumberOfTime30-59DaysPastDueNotWorse  150000 non-null  float64
 4   DebtRatio                             150000 non-null  float64
 5   MonthlyIncome                         150000 non-null  float64
 6   NumberOfOpenCreditLinesAndLoans       150000 non-null  float64
 7   NumberOfTimes90DaysLate               150000 non-null  float64
 8   NumberRealEstateLoansOrLines          150000 non-null  float64
 9   NumberOfTime60-89DaysPastDueNotWorse  150000 non-null  float64
 10  NumberOfDependents                    150000 non-null  float64
dtype

In [25]:
predictions_df_woe.head(10)

SeriousDlqin2yrs  RevolvingUtilizationOfUnsecuredLines       age  \
0                 1                             -0.796789  0.070228   
1                 0                             -0.796789 -0.311206   
2                 0                             -0.796789 -0.311206   
3                 0                              0.949764 -0.554284   
4                 0                             -0.796789  0.070228   
5                 0                              0.949764  1.063738   
6                 0                              0.949764  0.070228   
7                 0                             -0.796789 -0.311206   
8                 0                              0.949764 -0.554284   
9                 0                              0.949764  0.070228   

   NumberOfTime30-59DaysPastDueNotWorse  DebtRatio  MonthlyIncome  \
0                             -2.251409  -0.575353       0.111096   
1                              0.541721   0.175061      -0.311170   
2                             -1.091567   0.175061      -0.311170   
3                              0.541721   0.175061      -0.311170   
4                             -1.091567   0.175061       0.460433   
5                              0.541721  -0.067773      -0.311170   
6                              0.541721   0.193661       0.205179   
7                              0.541721   0.175061      -0.311170   
8                              0.541721   0.193661       0.205179   
9                              0.541721  -0.067773       0.460433   

   NumberOfOpenCreditLinesAndLoans  NumberOfTimes90DaysLate  \
0                         0.035635                 0.176674   
1                        -0.004249                 0.176674   
2                        -0.004249                 0.176674   
3                        -0.004249                 0.176674   
4                        -0.004249                 0.176674   
5                        -0.004249                 0.176674   
6                        -0.004249                 0.176674   
7                        -0.004249                 0.176674   
8                        -0.004249                 0.176674   
9                        -0.004249                 0.176674   

   NumberRealEstateLoansOrLines  NumberOfTime60-89DaysPastDueNotWorse  \
0                     -0.592980                              0.288208   
1                     -0.023090                              0.288208   
2                     -0.023090                              0.288208   
3                     -0.023090                              0.288208   
4                      0.188639                              0.288208   
5                      0.188639                              0.288208   
6                     -0.592980                              0.288208   
7                     -0.023090                              0.288208   
8                     -0.023090                              0.288208   
9                     -0.592980                              0.288208   

   NumberOfDependents  
0           -0.352272  
1            0.078749  
2            0.078749  
3            0.078749  
4            0.078749  
5            0.078749  
6            0.078749  
7            0.078749  
8           -0.126763  
9           -0.352272

##### Logistic Coefficients #####

In [26]:
logreg = LogisticRegression()

In [27]:
coefs = []
intercepts = []

In [28]:
for var in var_list:
    logreg.fit(predictions_df_woe[[var]], predictions_df_woe['SeriousDlqin2yrs']);
    coefs.append(logreg.coef_[0][0])
    intercepts.append(logreg.intercept_[0])

In [29]:
coef_df = pd.DataFrame(zip(var_list, coefs, intercepts), columns=['Variable', 'Coeffient', 'Intercept'])

In [30]:
coef_df

Variable  Coeffient  Intercept
0  RevolvingUtilizationOfUnsecuredLines  -1.025263  -2.534305
1                                   age  -0.983258  -2.618690
2  NumberOfTime30-59DaysPastDueNotWorse  -0.876607  -2.701037
3                             DebtRatio  -0.968957  -2.624434
4                         MonthlyIncome  -0.955729  -2.640271
5       NumberOfOpenCreditLinesAndLoans  -0.778155  -2.636247
6               NumberOfTimes90DaysLate  -0.996717  -2.636661
7          NumberRealEstateLoansOrLines  -0.765749  -2.603261
8  NumberOfTime60-89DaysPastDueNotWorse  -0.973950  -2.640623
9                    NumberOfDependents  -0.780107  -2.661430

### Logistic Regression on the full WoE dataset ###

In [47]:
logreg = LogisticRegression()

In [48]:
predictions_df_woe.columns

Index(['SeriousDlqin2yrs', 'RevolvingUtilizationOfUnsecuredLines', 'age',
       'NumberOfTime30-59DaysPastDueNotWorse', 'DebtRatio', 'MonthlyIncome',
       'NumberOfOpenCreditLinesAndLoans', 'NumberOfTimes90DaysLate',
       'NumberRealEstateLoansOrLines', 'NumberOfTime60-89DaysPastDueNotWorse',
       'NumberOfDependents'],
      dtype='object')

In [49]:
X = predictions_df_woe.drop(columns=['SeriousDlqin2yrs', 'NumberOfOpenCreditLinesAndLoans'])
y = predictions_df_woe['SeriousDlqin2yrs']

NumberOfOpenCreditLinesAndLoans is dropped because it has a very low IV of 0.000375

In [50]:
logreg.fit(X, y);

In [51]:
logreg.coef_[0]

array([-0.67505692, -0.47860843, -0.49756609, -0.5557159 , -0.36903108,
       -0.51641583, -0.77320443, -0.44242076, -0.20012398])

In [55]:
scorecard_coefs = pd.DataFrame(zip(X.columns, logreg.coef_[0]), columns=['variable', 'coefficient'])

In [56]:
scorecard_coefs

variable  coefficient
0  RevolvingUtilizationOfUnsecuredLines    -0.675057
1                                   age    -0.478608
2  NumberOfTime30-59DaysPastDueNotWorse    -0.497566
3                             DebtRatio    -0.555716
4                         MonthlyIncome    -0.369031
5               NumberOfTimes90DaysLate    -0.516416
6          NumberRealEstateLoansOrLines    -0.773204
7  NumberOfTime60-89DaysPastDueNotWorse    -0.442421
8                    NumberOfDependents    -0.200124

In [71]:
scorecard_coefs = scorecard_coefs.append(pd.DataFrame([['Intercept', logreg.intercept_[0]]], columns=['variable', 'coefficient']))

In [36]:
predictions_df_woe.to_csv('../../data/woe_data.csv', index=False)

In [72]:
scorecard_coefs.to_csv('../../data/scorecard_coefs.csv', index=False)